File: Graphs/GraphADT.java
// File Name: GraphADT.java

import java.util.*;

class GraphADT {

    private Map<Integer, List<Integer>> adj = new HashMap<>();

    public void insertVertex(int v) {
        adj.putIfAbsent(v, new ArrayList<>());
    }

    public void insertEdge(int u, int v) {
        adj.get(u).add(v);
        adj.get(v).add(u);
    }

    public void deleteEdge(int u, int v) {
        adj.get(u).remove(Integer.valueOf(v));
        adj.get(v).remove(Integer.valueOf(u));
    }

    public void deleteVertex(int v) {
        adj.remove(v);
        for (List<Integer> list : adj.values()) {
            list.remove(Integer.valueOf(v));
        }
    }

    public void display() {
        for (int v : adj.keySet()) {
            System.out.println(v + " -> " + adj.get(v));
        }
    }

    public void BFS(int start) {
        Set<Integer> visited = new HashSet<>();
        Queue<Integer> queue = new LinkedList<>();
        queue.add(start);
        visited.add(start);

        while (!queue.isEmpty()) {
            int v = queue.poll();
            System.out.print(v + " ");
            for (int n : adj.get(v)) {
                if (!visited.contains(n)) {
                    visited.add(n);
                    queue.add(n);
                }
            }
        }
    }

    public void DFS(int start, Set<Integer> visited) {
        visited.add(start);
        System.out.print(start + " ");
        for (int n : adj.get(start)) {
            if (!visited.contains(n)) {
                DFS(n, visited);
            }
        }
    }

    public static void main(String[] args) {
        GraphADT g = new GraphADT();
        g.insertVertex(0);
        g.insertVertex(1);
        g.insertVertex(2);

        g.insertEdge(0,1);
        g.insertEdge(1,2);

        System.out.println("Graph:");
        g.display();

        System.out.print("BFS: ");
        g.BFS(0);

        System.out.print("\nDFS: ");
        g.DFS(0, new HashSet<>());
    }
}

File: Graphs/GraphADT.py
# File Name: GraphADT.py

class GraphADT:
    def __init__(self):
        self.graph = {}

    def insert_vertex(self, v):
        if v not in self.graph:
            self.graph[v] = []

    def insert_edge(self, u, v):
        self.graph[u].append(v)
        self.graph[v].append(u)

    def delete_edge(self, u, v):
        if v in self.graph[u]:
            self.graph[u].remove(v)
        if u in self.graph[v]:
            self.graph[v].remove(u)

    def delete_vertex(self, v):
        if v in self.graph:
            for n in self.graph[v]:
                self.graph[n].remove(v)
            del self.graph[v]

    def display(self):
        for v in self.graph:
            print(v, "->", self.graph[v])

    def bfs(self, start):
        visited = set()
        queue = [start]
        visited.add(start)

        while queue:
            v = queue.pop(0)
            print(v, end=" ")
            for n in self.graph[v]:
                if n not in visited:
                    visited.add(n)
                    queue.append(n)

    def dfs(self, start, visited=None):
        if visited is None:
            visited = set()
        visited.add(start)
        print(start, end=" ")
        for n in self.graph[start]:
            if n not in visited:
                self.dfs(n, visited)

if __name__ == "__main__":
    g = GraphADT()
    g.insert_vertex(0)
    g.insert_vertex(1)
    g.insert_vertex(2)

    g.insert_edge(0,1)
    g.insert_edge(1,2)

    print("Graph:")
    g.display()

    print("BFS:", end=" ")
    g.bfs(0)

    print("\nDFS:", end=" ")
    g.dfs(0)
